stats.c

```diff
@@ -688,6 +716,142 @@ void update_checksum(bam1_t *bam_line, stats_t *stats)
+                if (!stats->nbases_cr) {
+                    stats->nbases_cr = barcode_len;
+
+                    stats->acgtno_cr = calloc(stats->nbases_cr, sizeof(acgtno_count_t));
+                    stats->quals_cr  = calloc(stats->nbases_cr*stats->nquals, sizeof(uint64_t));
+
+                    if (!stats->acgtno_cr || !stats->quals_cr)
+                        error("Error allocating memory. Aborting!\n");
+                }
+
+                nbases = stats->nbases_cr;
+                maxqual = &stats->maxqual_cr;
+                acgtno = stats->acgtno_cr;
+                quals = stats->quals_cr;
+                separator = &stats->separator_cr;
+            } else if (!strncmp(barcode_tag, "OX", 2)) {
+                if (!stats->nbases_ox) {
+                    stats->nbases_ox = barcode_len;
+
+                    stats->acgtno_ox = calloc(stats->nbases_ox, sizeof(acgtno_count_t));
+                    stats->quals_ox  = calloc(stats->nbases_ox*stats->nquals, sizeof(uint64_t));
+
+                    if (!stats->acgtno_ox || !stats->quals_ox)
+                        error("Error allocating memory. Aborting!\n");
+                }
+
+                nbases = stats->nbases_ox;
+                maxqual = &stats->maxqual_ox;
+                acgtno = stats->acgtno_ox;
+                quals = stats->quals_ox;
+                separator = &stats->separator_ox;
+            } else if (!strncmp(barcode_tag, "RX", 2)) {
+                if (!stats->nbases_rx) {
+                    stats->nbases_rx = barcode_len;
+
+                    stats->acgtno_rx = calloc(stats->nbases_rx, sizeof(acgtno_count_t));
+                    stats->quals_rx  = calloc(stats->nbases_rx*stats->nquals, sizeof(uint64_t));
+
+                    if (!stats->acgtno_rx || !stats->quals_rx)
+                        error("Error allocating memory. Aborting!\n");
+                }
+
+                nbases = stats->nbases_rx;
+                maxqual = &stats->maxqual_rx;
+                acgtno = stats->acgtno_rx;
+                quals = stats->quals_rx;
+                separator = &stats->separator_rx;
+            } else {
+                return 2;
+            }
+
+            if (barcode_len > nbases)
+                error("Barcodes differ in length at sequence '%s'. Aborting!\n", bam_get_qname(bam_line));
+
+            for (i = 0; i < barcode_len; i++) {
+                switch (barcode[i]) {
+                case 'A':
+                    acgtno[i].a++;
+                    break;
+                case 'C':
+                    acgtno[i].c++;
+                    break;
+                case 'G':
+                    acgtno[i].g++;
+                    break;
+                case 'T':
+                    acgtno[i].t++;
+                    break;
+                case 'N':
+                    acgtno[i].n++;
+                    break;
+                default:
+                    if (*separator > 0) {
```

Comment:
Separator is initialised to -1 and 0 can be a legal index (well, illegal probably, but it could be set to that with malformed data and the loop above would put it there).   So maybe this needs to be `if (*separator >= 0) {`.